### ISSUE TITLE
Support for Tiger-192 / haval192

### ISSUE THREAD

**Issue by bvernoux** [iagox86/hash_extender#6]
Could you add support of Tiger-192 / haval192 which seems to be affected by this attack ?

**Comment by iagox86** [comment]
I did a bit of research. It looks like Tiger-192 should work for sure, but might be a little tricky: it uses slightly different padding. I don't remember how hash_extender works, or if that'd be a problem. FWIW, Tiger-128 and Tiger-160 appear to be truncations, so they wouldn't be vulnerable.

I had trouble finding any information about HAVAL from a quick Google search, and Wikipedia doesn't mention if it's Merkle-Damgård or not, so I don't know if it's vulnerable. I'll have to find a whitepaper or something.

I have a lot on my plate for the next couple months, but I put this on my list. I don't think I'll get around to it till probably spring, though, so if anybody else wants to take this, it shouldn't be super crazy. :)

**Comment by nanoeve** [comment]
The HAVAL paper is available with DOI [10.1007/3-540-57220-1_54](https://doi.org/10.1007/3-540-57220-1_54).

I do not think that HAVAL-192 is affected by the attack since it is a truncation.

```c
#elif FPTLEN == 192
  temp = (state->fingerprint[7] &  (haval_word)0x1F) | 
         (state->fingerprint[6] & ((haval_word)0x3F << 26));
  state->fingerprint[0] += rotate_right(temp, 26);

  temp = (state->fingerprint[7] & ((haval_word)0x1F <<  5)) | 
         (state->fingerprint[6] &  (haval_word)0x1F);
  state->fingerprint[1] += temp;

  temp = (state->fingerprint[7] & ((haval_word)0x3F << 10)) | 
         (state->fingerprint[6] & ((haval_word)0x1F <<  5));
  state->fingerprint[2] += temp >> 5;

  temp = (state->fingerprint[7] & ((haval_word)0x1F << 16)) | 
         (state->fingerprint[6] & ((haval_word)0x3F << 10));
  state->fingerprint[3] += temp >> 10;

  temp = (state->fingerprint[7] & ((haval_word)0x1F << 21)) | 
         (state->fingerprint[6] & ((haval_word)0x1F << 16));
  state->fingerprint[4] += temp >> 16;

  temp = (state->fingerprint[7] & ((haval_word)0x3F << 26)) | 
         (state->fingerprint[6] & ((haval_word)0x1F << 21));
  state->fingerprint[5] += temp >> 21;
```
